File: fbp/flow.py
"""Lane-level item flow: what can be on each lane of every belt.

This is a static "could be here" analysis, not a simulation. Item sets only
grow along the graph and the computation runs to a fixed point with a
worklist, so it is linear in the size of the print rather than in the length
of its longest belt:

  * an inserter dropping onto a belt puts its source's items on the FAR lane
  * belts keep lanes through straights and curves
  * a belt entering the SIDE of another belt sideloads: both of its lanes land
    on the target's NEAR lane (a curve is a single input with nothing behind)
  * a sideload into an underground tile passes only the feeder lane aligned
    with the open half: the back half for an entrance, the front half for an
    exit (docs/belt-patterns.md section 1)
  * splitters keep lanes and send everything to both outputs, unless a filter
    is set: then the filtered item goes only to the priority output and the
    rest only to the other one
  * a crafter emits its recipe's item products; a furnace emits the smelting
    result of whatever reaches it (or "smelted?" when its input is unknown)
  * a chest holds whatever is dropped into it; inserters picking from it carry
    that on

Edge feeders let the user declare items entering from outside the print:
[{"x": 23, "y": 86, "items": ["iron-ore"], "lane": "both"}], stored beside
the blueprint as <name>.feeds.json.
"""
import json
import os
from collections import defaultdict, deque

from .model import DIRS, INSERTERS, inserter_ends, is_splitter, is_underground, kind, tile_of
from .trace import Tracer

UNKNOWN_SMELT = "smelted?"


def feeds_path(blueprint_path):
    stem, _ = os.path.splitext(blueprint_path)
    return stem + ".feeds.json"


def load_feeds(path):
    if path and os.path.exists(path):
        with open(path, encoding="utf-8") as handle:
            return json.load(handle)
    return []


def save_feeds(path, feeds):
    with open(path, "w", encoding="utf-8") as handle:
        json.dump(feeds, handle, indent=1)


def left_of(direction):
    dx, dy = DIRS[direction & 12]
    return (dy, -dx)


class Flow:
    def __init__(self, grid, gamedata, feeds=()):
        self.g = grid
        self.gd = gamedata
        self.tr = Tracer(grid, gamedata)
        self.feeds = list(feeds)
        self.lanes = {}        # belt id -> {"left": set, "right": set}
        self.machine_in = {}   # machine id -> set
        self.machine_out = {}  # machine id -> set
        self.chest = {}        # chest id -> set
        self.smelt = self._smelting_map()
        self._build_graph()

    def _smelting_map(self):
        out = {}
        for name, r in self.gd.recipes.items():
            if r.get("category") == "smelting":
                for ing in r["ingredients"]:
                    out.setdefault(ing, set()).update(r["results"])
        return out

    # -- static graph -------------------------------------------------------
    def _build_graph(self):
        g, tr = self.g, self.tr
        self.belt_in = defaultdict(list)      # belt id -> [(feeder, cell)]
        self.belt_out = defaultdict(set)      # belt id -> {downstream belt ids}
        for b in g.entities:
            if kind(b) != "belt":
                continue
            for cell in g.cells_of(b):
                for f in tr.upstream(b):
                    if cell in tr.outputs(f):
                        self.belt_in[b["entity_number"]].append((f, cell))
                        self.belt_out[f["entity_number"]].add(b["entity_number"])
        self.ins_pick = {}                    # inserter id -> [source entities]
        self.ins_drop = {}                    # inserter id -> [(target entity, drop cell)]
        self.picked_by = defaultdict(list)    # source id -> [inserter]
        self.machine_ins = defaultdict(list)  # machine id -> [inserter dropping into it]
        for ins in g.entities:
            if ins["name"] not in INSERTERS:
                continue
            pick, drop = inserter_ends(ins)
            srcs = g.at(*pick, exclude_kinds=("inserter",))
            tgts = g.at(*drop, exclude_kinds=("inserter",))
            self.ins_pick[ins["entity_number"]] = srcs
            self.ins_drop[ins["entity_number"]] = [(t, drop) for t in tgts]
            for s in srcs:
                self.picked_by[s["entity_number"]].append(ins)
            for t in tgts:
                if kind(t) in ("crafter", "furnace"):
                    self.machine_ins[t["entity_number"]].append(ins)

    # -- lane helpers --------------------------------------------------------
    def lanes_of(self, b):
        return self.lanes.setdefault(b["entity_number"], {"left": set(), "right": set()})

    def _feeder_side(self, f, b_dir, cell):
        fd = f.get("direction", 0) & 12
        if fd == (b_dir & 12):
            return "behind"
        if is_splitter(f):
            dx, dy = DIRS[fd]
            fx, fy = next((c for c in self.g.cells_of(f) if (c[0] + dx, c[1] + dy) == cell), self.g.cells_of(f)[0])
        else:
            fx, fy = tile_of(f)
        rel = (fx - cell[0], fy - cell[1])
        return "left" if rel == left_of(b_dir) else "right"

    def _splitter_sides(self, s):
        cells = self.g.cells_of(s)
        lv = left_of(s.get("direction", 0))
        return sorted(cells, key=lambda c: -(c[0] * lv[0] + c[1] * lv[1]))

    def _out_lanes(self, f, cell):
        """Lane sets leaving feeder f toward `cell` (splitter filters applied here)."""
        FL = self.lanes_of(f)
        if not is_splitter(f):
            return {"left": FL["left"], "right": FL["right"]}
        flt = (f.get("filter") or {}).get("name")
        prio = f.get("output_priority")
        if not flt or prio not in ("left", "right"):
            return {"left": FL["left"], "right": FL["right"]}
        left_cell, right_cell = self._splitter_sides(f)
        dx, dy = DIRS[f.get("direction", 0) & 12]
        out_side = "left" if (left_cell[0] + dx, left_cell[1] + dy) == cell else "right"
        keep = (lambda s: {i for i in s if i == flt}) if out_side == prio else (lambda s: {i for i in s if i != flt})
        return {"left": keep(FL["left"]), "right": keep(FL["right"])}

    def _underground_filter(self, f, u, FL):
        ud = u.get("direction", 0) & 12
        fwd = DIRS[ud]
        back = (-fwd[0], -fwd[1])
        open_side = back if u.get("type") == "input" else fwd
        return FL["left"] if left_of(f.get("direction", 0)) == open_side else FL["right"]

    def _drop_lane(self, ins, belt, drop_cell):
        ix, iy = tile_of(ins)
        lv = left_of(belt.get("direction", 0))
        rel = (ix - drop_cell[0], iy - drop_cell[1])
        if rel == lv:
            return "right"       # standing on the belt's left: items land on the far (right) lane
        if rel == (-lv[0], -lv[1]):
            return "left"
        return "right"           # inline: the game uses the belt's right lane

    def _source_items(self, sources):
        items = set()
        for s in sources:
            k = kind(s)
            if k == "belt":
                L = self.lanes_of(s); items |= L["left"] | L["right"]
            elif k in ("crafter", "furnace"):
                items |= self.machine_out.get(s["entity_number"], set())
            elif k == "chest":
                items |= self.chest.get(s["entity_number"], set())
                for sec in s.get("request_filters", {}).get("sections", []):
                    items |= {flt["name"] for flt in sec.get("filters", []) if flt.get("name")}
        return items

    # -- worklist fixed point -----------------------------------------------
    def run(self):
        """Fixed point, repeated so that "smelted?" placeholders do not survive on furnaces whose
        input turned out to be known: each pass records which furnaces resolved, the next pass
        starts clean and lets those furnaces emit only real plates. Converges in 2-3 passes."""
        self._resolved = set()
        while True:
            self.lanes.clear(); self.machine_in.clear(); self.machine_out.clear(); self.chest.clear()
            self._run_once()
            resolved = {eid for eid, mi in self.machine_in.items()
                        if kind(self.g.byid[eid]) == "furnace" and any(i in self.smelt for i in mi)}
            if resolved <= self._resolved:
                return self
            self._resolved |= resolved

    def _run_once(self):
        g = self.g
        work = deque()
        queued = set()

        def push(eid):
            if eid not in queued:
                queued.add(eid); work.append(eid)

        def add_lane(b, lane, items):
            L = self.lanes_of(b)
            new = items - L[lane]
            if new:
                L[lane] |= new
                bid = b["entity_number"]
                for d in self.belt_out[bid]:
                    push(d)
                for ins in self.picked_by[bid]:
                    push(ins["entity_number"])

        for f in self.feeds:
            b = g.belt_at(int(f["x"]), int(f["y"]))
            if b:
                for lane in (("left", "right") if f.get("lane", "both") == "both" else (f["lane"],)):
                    add_lane(b, lane, set(f.get("items", [])))
        for m in g.entities:                      # machines emit products even before inputs are known
            if kind(m) in ("crafter", "furnace"):
                push(m["entity_number"])

        while work:
            eid = work.popleft(); queued.discard(eid)
            e = g.byid.get(eid)
            if e is None:
                continue
            k = kind(e)
            if k == "belt":
                d = e.get("direction", 0)
                for cell in g.cells_of(e):
                    feeders = [(f, c) for f, c in self.belt_in[eid] if c == cell]
                    if not feeders:
                        continue
                    behind = [f for f, c in feeders if self._feeder_side(f, d, cell) == "behind"]
                    sides = [f for f, c in feeders if f not in behind]
                    curve = not behind and len(sides) == 1 and not is_underground(e) and not is_splitter(e)
                    for f in behind:
                        FL = self._out_lanes(f, cell)
                        add_lane(e, "left", FL["left"]); add_lane(e, "right", FL["right"])
                    for f in sides:
                        FL = self._out_lanes(f, cell)
                        if curve:
                            add_lane(e, "left", FL["left"]); add_lane(e, "right", FL["right"])
                        else:
                            incoming = self._underground_filter(f, e, FL) if is_underground(e) else (FL["left"] | FL["right"])
                            add_lane(e, self._feeder_side(f, d, cell), incoming)
            elif e["name"] in INSERTERS:
                items = self._source_items(self.ins_pick.get(eid, []))
                if not items:
                    continue
                for t, drop in self.ins_drop.get(eid, []):
                    tk = kind(t)
                    if tk == "belt":
                        add_lane(t, self._drop_lane(e, t, drop), items)
                    elif tk == "chest":
                        c = self.chest.setdefault(t["entity_number"], set())
                        new = items - c
                        if new:
                            c |= new
                            for ins in self.picked_by[t["entity_number"]]:
                                push(ins["entity_number"])
                    elif tk in ("crafter", "furnace"):
                        push(t["entity_number"])
            elif k in ("crafter", "furnace"):
                mi = self.machine_in.setdefault(eid, set())
                for ins in self.machine_ins[eid]:
                    mi |= self._source_items(self.ins_pick.get(ins["entity_number"], []))
                if k == "crafter":
                    out = set(self.gd.products(e["recipe"])) if e.get("recipe") else set()
                else:
                    out = set()
                    for i in mi:
                        out |= self.smelt.get(i, set())
                    if not out and eid not in self._resolved:
                        out = {UNKNOWN_SMELT}
                mo = self.machine_out.setdefault(eid, set())
                new = out - mo
                if new:
                    mo |= new
                    for ins in self.picked_by[eid]:
                        push(ins["entity_number"])
        return self

    # -- results ----------------------------------------------------------
    def machine_status(self, m):
        """('ok'|'missing'|'unknown'|'none', missing_items). 'unknown' when a furnace of
        unresolved input could be supplying the missing plates; 'none' for no recipe."""
        k = kind(m)
        mi = self.machine_in.get(m["entity_number"], set())
        if k == "furnace":
            known = mi - {UNKNOWN_SMELT}
            if known:
                return ("ok", []) if any(i in self.smelt for i in known) else ("missing", sorted(known))
            return ("unknown", [])
        if not m.get("recipe"):
            return ("none", [])
        needs = self.gd.ingredients(m["recipe"])
        if needs is None:
            return ("unknown", [])
        missing = [i for i in needs if i not in mi]
        if not missing:
            return ("ok", [])
        plates = set().union(*self.smelt.values()) if self.smelt else set()
        if UNKNOWN_SMELT in mi and all(i in plates for i in missing):
            return ("unknown", missing)
        return ("missing", missing)

    def items_seen(self):
        seen = set()
        for L in self.lanes.values():
            seen |= L["left"] | L["right"]
        return seen

    def as_json(self):
        machines = {}
        for m in self.g.entities:
            if kind(m) in ("crafter", "furnace"):
                status, missing = self.machine_status(m)
                machines[str(m["entity_number"])] = {"in": sorted(self.machine_in.get(m["entity_number"], ())),
                                                     "out": sorted(self.machine_out.get(m["entity_number"], ())),
                                                     "status": status, "missing": missing}
        return {"lanes": {str(k): {"left": sorted(v["left"]), "right": sorted(v["right"])}
                          for k, v in self.lanes.items() if v["left"] or v["right"]},
                "machines": machines,
                "chests": {str(k): sorted(v) for k, v in self.chest.items()},
                "feeds": self.feeds,
                "items": sorted(self.items_seen())}

    def at(self, x, y):
        b = self.g.belt_at(x, y)
        if not b:
            return None
        return self.lanes_of(b)

File: vscode/media/flow.js
// Lane-level item flow, JavaScript port of fbp/flow.py. Same rules, same worklist, same
// output shape, so the page can compute it without the extension host. Runs in the browser
// (window.FBPFlow) and in Node (module.exports) so tests can compare it with the Python.
(function (root) {
  const DIRS = { 0: [0, -1], 4: [1, 0], 8: [0, 1], 12: [-1, 0] };
  const UNKNOWN = "smelted?";
  const REACH = { "underground-belt": 5, "fast-underground-belt": 7, "express-underground-belt": 9, "turbo-underground-belt": 11 };
  const INSERTERS = new Set(["inserter", "fast-inserter", "long-handed-inserter", "bulk-inserter", "stack-inserter", "burner-inserter"]);
  const CRAFTERS = new Set(["assembling-machine-1", "assembling-machine-2", "assembling-machine-3", "chemical-plant", "oil-refinery",
    "centrifuge", "electromagnetic-plant", "foundry", "biochamber", "cryogenic-plant", "recycler"]);
  const FURNACES = new Set(["stone-furnace", "steel-furnace", "electric-furnace"]);
  const CHESTS = new Set(["wooden-chest", "iron-chest", "steel-chest", "passive-provider-chest", "active-provider-chest",
    "storage-chest", "buffer-chest", "requester-chest"]);
  const BUILTIN = {
    "assembling-machine-1": [3, 3], "assembling-machine-2": [3, 3], "assembling-machine-3": [3, 3], "chemical-plant": [3, 3],
    "oil-refinery": [5, 5], "centrifuge": [3, 3], "lab": [3, 3], "beacon": [3, 3], "roboport": [4, 4], "radar": [3, 3],
    "stone-furnace": [2, 2], "steel-furnace": [2, 2], "electric-furnace": [3, 3], "storage-tank": [3, 3], "pumpjack": [3, 3],
    "electric-mining-drill": [3, 3], "big-electric-pole": [2, 2], "substation": [2, 2], "accumulator": [2, 2], "solar-panel": [3, 3],
    "splitter": [2, 1], "fast-splitter": [2, 1], "express-splitter": [2, 1], "turbo-splitter": [2, 1], "train-stop": [2, 2],
  };
  const UNMODELLED = new Set(["legacy-straight-rail", "legacy-curved-rail", "straight-rail", "curved-rail-a", "curved-rail-b",
    "half-diagonal-rail", "cargo-wagon", "locomotive", "fluid-wagon", "artillery-wagon"]);

  const isBelt = (n) => /transport-belt$|underground-belt$|splitter$/.test(n);
  const isUnderground = (n) => n.endsWith("underground-belt");
  const isSplitter = (n) => n.endsWith("splitter");
  function kind(e) {
    const n = e.name;
    if (isBelt(n)) return "belt";
    if (INSERTERS.has(n)) return "inserter";
    if (CRAFTERS.has(n)) return "crafter";
    if (FURNACES.has(n)) return "furnace";
    if (CHESTS.has(n)) return "chest";
    return "other";
  }
  const leftOf = (d) => { const [dx, dy] = DIRS[d & 12] || [0, -1]; return [dy, -dx]; };
  const key = (x, y) => x + "," + y;
  const same = (a, b) => a[0] === b[0] && a[1] === b[1];

  function cellsOf(e, footprints) {
    if (UNMODELLED.has(e.name)) return [[Math.floor(e.position.x), Math.floor(e.position.y)]];
    let [w, h] = (footprints && footprints[e.name]) || BUILTIN[e.name] || [1, 1];
    const d = (e.direction || 0) & 12;
    if (d === 4 || d === 12) [w, h] = [h, w];
    const x0 = Math.floor(e.position.x - w / 2), y0 = Math.floor(e.position.y - h / 2);
    const out = [];
    for (let dx = 0; dx < w; dx++) for (let dy = 0; dy < h; dy++) out.push([x0 + dx, y0 + dy]);
    return out;
  }
  const tileOf = (e) => [Math.floor(e.position.x), Math.floor(e.position.y)];
  function inserterEnds(e) {
    const [dx, dy] = DIRS[(e.direction || 0) & 12];
    const r = e.name === "long-handed-inserter" ? 2 : 1;
    const [x, y] = tileOf(e);
    return [[x + dx * r, y + dy * r], [x - dx * r, y - dy * r]];
  }

  // ---------------------------------------------------------------- game data
  function makeData(gd) {
    const recipes = (gd && gd.recipes) || {};
    const smelt = {};
    for (const r of Object.values(recipes)) {
      if (r.category !== "smelting") continue;
      for (const ing of Object.keys(r.ingredients || {})) { smelt[ing] = smelt[ing] || new Set(); for (const p of Object.keys(r.results || {})) smelt[ing].add(p); }
    }
    const plates = new Set(); for (const s of Object.values(smelt)) for (const p of s) plates.add(p);
    return {
      recipes, smelt, plates,
      products(name) {
        const r = recipes[name]; if (!r) return new Set([name]);
        const fl = new Set(r.fluid_results || []);
        const out = new Set(Object.keys(r.results || {}).filter((k) => !fl.has(k)));
        return out.size || fl.size ? out : new Set([name]);
      },
      ingredients(name) {
        const r = recipes[name]; if (!r) return null;
        const fl = new Set(r.fluid_ingredients || []);
        return Object.keys(r.ingredients || {}).filter((k) => !fl.has(k));
      },
    };
  }

  // ---------------------------------------------------------------- grid + belt graph
  function buildGrid(entities, footprints) {
    const byid = new Map(), tiles = new Map(), cells = new Map();
    for (const e of entities) {
      byid.set(e.entity_number, e);
      const cs = cellsOf(e, footprints); cells.set(e.entity_number, cs);
      for (const c of cs) { const k = key(c[0], c[1]); if (!tiles.has(k)) tiles.set(k, []); tiles.get(k).push(e); }
    }
    const at = (x, y, noInserters) => (tiles.get(key(x, y)) || []).filter((e) => !noInserters || !INSERTERS.has(e.name));
    const beltAt = (x, y) => (tiles.get(key(x, y)) || []).find((e) => isBelt(e.name)) || null;
    return { entities, byid, tiles, cells, at, beltAt, cellsOf: (e) => cells.get(e.entity_number) };
  }

  function ugPair(g, b) {
    const d = (b.direction || 0) & 12, [dx, dy] = DIRS[d];
    const want = b.type === "input" ? "output" : "input", step = b.type === "input" ? 1 : -1;
    const [x, y] = tileOf(b), reach = REACH[b.name] || 5;
    for (let k = 1; k <= reach; k++) {
      for (const f of g.at(x + dx * k * step, y + dy * k * step)) {
        if (f.name === b.name && f.type === want && ((f.direction || 0) & 12) === d) return f;
      }
    }
    return null;
  }
  function outputs(g, b) {
    const d = (b.direction || 0) & 12, [dx, dy] = DIRS[d];
    if (isSplitter(b.name)) return g.cellsOf(b).map((c) => [c[0] + dx, c[1] + dy]);
    const [x, y] = tileOf(b);
    if (isUnderground(b.name) && b.type === "input") { const p = ugPair(g, b); return p ? [tileOf(p)] : []; }
    return [[x + dx, y + dy]];
  }
  // An inserter drops on the FAR lane: the side of the belt away from where it stands. An inserter
  // in line with the belt has no far side; the game puts those on the right. The viewer draws its
  // hover highlight from this, so the picture and the flow can never disagree.
  function dropLane(ins, belt, dropCell) {
    const [ix, iy] = tileOf(ins), lv = leftOf(belt.direction || 0);
    const rel = [Math.sign(ix - dropCell[0]), Math.sign(iy - dropCell[1])];   // long-handed inserters stand 2 tiles away
    if (same(rel, lv)) return "right";
    if (same(rel, [-lv[0], -lv[1]])) return "left";
    return "right";
  }
  const sameAxis = (a, b) => (((a.direction || 0) & 12) % 8) === (((b.direction || 0) & 12) % 8);
  const opposed = (a, b) => ((((a.direction || 0) & 12) + 8) % 16) === ((b.direction || 0) & 12);
  function upstream(g, b) {
    // an underground exit receives through its tunnel and from belts running into its SIDE; the
    // flow's underground filter then keeps only the lane aligned with the open half. Two belts
    // facing each other never connect: head-on into the hood of an entrance, or into the front
    // of any belt, the items just pile up at the end.
    const found = new Map();
    const exit = isUnderground(b.name) && b.type === "output";
    if (exit) { const p = ugPair(g, b); if (p) found.set(p.entity_number, p); }
    for (const [x, y] of g.cellsOf(b)) {
      for (const [dx, dy] of Object.values(DIRS)) {
        const f = g.beltAt(x + dx, y + dy);
        if (!f || f === b) continue;
        if (exit && sameAxis(f, b)) continue;
        if (opposed(f, b)) continue;
        if (outputs(g, f).some((o) => same(o, [x, y]))) found.set(f.entity_number, f);
      }
    }
    return [...found.values()];
  }

  // ---------------------------------------------------------------- flow
  function compute(bp, footprints, gamedata, feeds) {
    const gd = makeData(gamedata);
    const g = buildGrid(bp.entities || [], footprints);
    feeds = feeds || [];

    // static graph
    const beltIn = new Map(), beltOut = new Map(), insPick = new Map(), insDrop = new Map(), pickedBy = new Map(), machineIns = new Map();
    const push2 = (m, k, v) => { if (!m.has(k)) m.set(k, []); m.get(k).push(v); };
    for (const b of g.entities) {
      if (!isBelt(b.name)) continue;
      for (const cell of g.cellsOf(b)) {
        for (const f of upstream(g, b)) {
          if (outputs(g, f).some((o) => same(o, cell))) {
            push2(beltIn, b.entity_number, [f, cell]);
            if (!beltOut.has(f.entity_number)) beltOut.set(f.entity_number, new Set());
            beltOut.get(f.entity_number).add(b.entity_number);
          }
        }
      }
    }
    for (const ins of g.entities) {
      if (!INSERTERS.has(ins.name)) continue;
      const [pick, drop] = inserterEnds(ins);
      const srcs = g.at(pick[0], pick[1], true), tgts = g.at(drop[0], drop[1], true);
      insPick.set(ins.entity_number, srcs); insDrop.set(ins.entity_number, tgts.map((t) => [t, drop]));
      for (const s of srcs) push2(pickedBy, s.entity_number, ins);
      for (const t of tgts) { const k = kind(t); if (k === "crafter" || k === "furnace") push2(machineIns, t.entity_number, ins); }
    }

    const lanes = new Map(), machineIn = new Map(), machineOut = new Map(), chest = new Map();
    const lanesOf = (b) => { if (!lanes.has(b.entity_number)) lanes.set(b.entity_number, { left: new Set(), right: new Set() }); return lanes.get(b.entity_number); };
    const union = (a, b) => { let changed = false; for (const x of b) if (!a.has(x)) { a.add(x); changed = true; } return changed; };

    function feederSide(f, bDir, cell) {
      const fd = (f.direction || 0) & 12;
      if (fd === (bDir & 12)) return "behind";
      let fx, fy;
      if (isSplitter(f.name)) {
        const [dx, dy] = DIRS[fd];
        const c = g.cellsOf(f).find((c) => same([c[0] + dx, c[1] + dy], cell)) || g.cellsOf(f)[0];
        [fx, fy] = c;
      } else [fx, fy] = tileOf(f);
      const lv = leftOf(bDir);
      return (fx - cell[0] === lv[0] && fy - cell[1] === lv[1]) ? "left" : "right";
    }
    function splitterSides(s) {
      const lv = leftOf(s.direction || 0);
      return g.cellsOf(s).slice().sort((a, b) => (b[0] * lv[0] + b[1] * lv[1]) - (a[0] * lv[0] + a[1] * lv[1]));
    }
    function outLanes(f, cell) {
      const FL = lanesOf(f);
      if (!isSplitter(f.name)) return FL;
      const flt = f.filter && f.filter.name, prio = f.output_priority;
      if (!flt || (prio !== "left" && prio !== "right")) return FL;
      const [leftCell] = splitterSides(f), [dx, dy] = DIRS[(f.direction || 0) & 12];
      const outSide = same([leftCell[0] + dx, leftCell[1] + dy], cell) ? "left" : "right";
      const keep = outSide === prio ? (s) => new Set([...s].filter((i) => i === flt)) : (s) => new Set([...s].filter((i) => i !== flt));
      return { left: keep(FL.left), right: keep(FL.right) };
    }
    function undergroundFilter(f, u, FL) {
      const fwd = DIRS[(u.direction || 0) & 12], back = [-fwd[0], -fwd[1]];
      const open = u.type === "input" ? back : fwd;
      return same(leftOf(f.direction || 0), open) ? FL.left : FL.right;
    }
    function sourceItems(sources) {
      const items = new Set();
      for (const s of sources) {
        const k = kind(s);
        if (k === "belt") { const L = lanesOf(s); for (const i of L.left) items.add(i); for (const i of L.right) items.add(i); }
        else if (k === "crafter" || k === "furnace") for (const i of machineOut.get(s.entity_number) || []) items.add(i);
        else if (k === "chest") {
          for (const i of chest.get(s.entity_number) || []) items.add(i);
          for (const sec of (s.request_filters || {}).sections || []) for (const f of sec.filters || []) if (f.name) items.add(f.name);
        }
      }
      return items;
    }

    let resolved = new Set();
    function runOnce() {
      lanes.clear(); machineIn.clear(); machineOut.clear(); chest.clear();
      const work = [], queued = new Set();
      const push = (id) => { if (!queued.has(id)) { queued.add(id); work.push(id); } };
      const addLane = (b, lane, items) => {
        if (union(lanesOf(b)[lane], items)) {
          for (const d of beltOut.get(b.entity_number) || []) push(d);
          for (const ins of pickedBy.get(b.entity_number) || []) push(ins.entity_number);
        }
      };
      for (const f of feeds) {
        const b = g.beltAt(Math.floor(f.x), Math.floor(f.y)); if (!b) continue;
        for (const lane of (f.lane || "both") === "both" ? ["left", "right"] : [f.lane]) addLane(b, lane, new Set(f.items || []));
      }
      for (const m of g.entities) { const k = kind(m); if (k === "crafter" || k === "furnace") push(m.entity_number); }
      let head = 0;
      while (head < work.length) {
        const eid = work[head++]; queued.delete(eid);
        const e = g.byid.get(eid); if (!e) continue;
        const k = kind(e);
        if (k === "belt") {
          const d = e.direction || 0;
          for (const cell of g.cellsOf(e)) {
            const feeders = (beltIn.get(eid) || []).filter(([f, c]) => same(c, cell)).map(([f]) => f);
            if (!feeders.length) continue;
            const behind = feeders.filter((f) => feederSide(f, d, cell) === "behind");
            const sides = feeders.filter((f) => !behind.includes(f));
            const curve = !behind.length && sides.length === 1 && !isUnderground(e.name) && !isSplitter(e.name);
            for (const f of behind) { const FL = outLanes(f, cell); addLane(e, "left", FL.left); addLane(e, "right", FL.right); }
            for (const f of sides) {
              const FL = outLanes(f, cell);
              if (curve) { addLane(e, "left", FL.left); addLane(e, "right", FL.right); }
              else {
                const incoming = isUnderground(e.name) ? undergroundFilter(f, e, FL) : new Set([...FL.left, ...FL.right]);
                addLane(e, feederSide(f, d, cell), incoming);
              }
            }
          }
        } else if (k === "inserter") {
          const items = sourceItems(insPick.get(eid) || []); if (!items.size) continue;
          for (const [t, drop] of insDrop.get(eid) || []) {
            const tk = kind(t);
            if (tk === "belt") addLane(t, dropLane(e, t, drop), items);
            else if (tk === "chest") {
              if (!chest.has(t.entity_number)) chest.set(t.entity_number, new Set());
              if (union(chest.get(t.entity_number), items)) for (const ins of pickedBy.get(t.entity_number) || []) push(ins.entity_number);
            } else if (tk === "crafter" || tk === "furnace") push(t.entity_number);
          }
        } else if (k === "crafter" || k === "furnace") {
          if (!machineIn.has(eid)) machineIn.set(eid, new Set());
          const mi = machineIn.get(eid);
          for (const ins of machineIns.get(eid) || []) union(mi, sourceItems(insPick.get(ins.entity_number) || []));
          let out;
          if (k === "crafter") out = e.recipe ? gd.products(e.recipe) : new Set();
          else {
            out = new Set();
            for (const i of mi) for (const p of gd.smelt[i] || []) out.add(p);
            if (!out.size && !resolved.has(eid)) out = new Set([UNKNOWN]);
          }
          if (!machineOut.has(eid)) machineOut.set(eid, new Set());
          if (union(machineOut.get(eid), out)) for (const ins of pickedBy.get(eid) || []) push(ins.entity_number);
        }
      }
    }
    for (let pass = 0; pass < 6; pass++) {
      runOnce();
      const now = new Set();
      for (const [eid, mi] of machineIn) { const e = g.byid.get(eid); if (kind(e) === "furnace" && [...mi].some((i) => gd.smelt[i])) now.add(eid); }
      if ([...now].every((x) => resolved.has(x))) break;
      for (const x of now) resolved.add(x);
    }

    function status(m) {
      const k = kind(m), mi = machineIn.get(m.entity_number) || new Set();
      if (k === "furnace") {
        const known = [...mi].filter((i) => i !== UNKNOWN);
        if (known.length) return known.some((i) => gd.smelt[i]) ? ["ok", []] : ["missing", known.sort()];
        return ["unknown", []];
      }
      if (!m.recipe) return ["none", []];
      const needs = gd.ingredients(m.recipe); if (needs === null) return ["unknown", []];
      const missing = needs.filter((i) => !mi.has(i));
      if (!missing.length) return ["ok", []];
      if (mi.has(UNKNOWN) && missing.every((i) => gd.plates.has(i))) return ["unknown", missing];
      return ["missing", missing];
    }

    const out = { lanes: {}, machines: {}, chests: {}, feeds, items: [] };
    const seen = new Set();
    for (const [id, L] of lanes) {
      if (!L.left.size && !L.right.size) continue;
      out.lanes[id] = { left: [...L.left].sort(), right: [...L.right].sort() };
      for (const i of L.left) seen.add(i); for (const i of L.right) seen.add(i);
    }
    for (const m of g.entities) {
      const k = kind(m); if (k !== "crafter" && k !== "furnace") continue;
      const [st, missing] = status(m);
      out.machines[m.entity_number] = { in: [...(machineIn.get(m.entity_number) || [])].sort(), out: [...(machineOut.get(m.entity_number) || [])].sort(), status: st, missing };
    }
    for (const [id, c] of chest) out.chests[id] = [...c].sort();
    out.items = [...seen].sort();
    return out;
  }

  const api = { compute, cellsOf, kind, dropLane, UNKNOWN };
  if (typeof module !== "undefined" && module.exports) module.exports = api;
  root.FBPFlow = api;
})(typeof window !== "undefined" ? window : globalThis);
